### PR TITLE
Add kmeans example

### PR DESCRIPTION
**Pull request by rylandg**
This is a draft PR.

That being said, I think the code is pretty polished. I expect to get some feedback in regards to details but I've tested all the flows/path and they work well.

What I'm not sure about is the README. For this initial pass I kept it minimal to avoid unnecessary extra work. Please let me know what areas need to be expanded/added and I will do so.

Thanks.

### CODE REVIEW

File: kmeans-sklearn/README.md [hunk 1]
```diff
@@ -0,0 +1,43 @@
+# KMeans via sklearn (Python2)
+
+Two clustering examples using [sklearn](https://scikit-learn.org/stable/)
```
Comment (avnerbraverman) [inline kmeans-sklearn/README.md]: Period at end of sentence

File: kmeans-sklearn/README.md [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+It is assumed you already have a Binaris account. If you don't have an account yet, worry not, signing up is painless and takes just 2 minutes. Visit [Getting Started](https://dev.binaris.com/tutorials/python/getting-started/) to find concise instructions for the process.
```
Comment (avnerbraverman) [inline kmeans-sklearn/README.md]: Maybe something shorter:

If this is your first time using Binaris, visit our [Getting Started] page to set up your account for free.

File: kmeans-sklearn/README.md [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
```
Comment (avnerbraverman) [inline kmeans-sklearn/README.md]: I think we definitely need a paragraph describing what we want to do: we will use kmeans to compress an imgae...

File: kmeans-sklearn/README.md [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    /path/to/bob/build.sh 2
+    ```
+
+    > Note: The first run of Bob builds a Docker image and may take a while
+
+3. Deploy your functions
+
+    ```bash
+    bn deploy public_kmeans_image_quant_py2
+    # or
```
Comment (avnerbraverman) [inline kmeans-sklearn/README.md]: Why 'or'?

File: kmeans-sklearn/README.md [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+    bn deploy public_kmeans_simple_py2
+    ```
+
+As a final note, keep in mind that you only need to run Bob again if you add or remove pip packages. Changes to the function code itself only require another `bn deploy`.
+
+## Try it out
+
+Visit the URL printed by deploy (for `public_kmeans_image_quant_py2`) in your browser and add the image URL you want to run kmeans on. The optional query param `num_colors` can also be used to modify the number of clusters kmeans used to quantify your image.
+
+_Example URL_
```
Comment (avnerbraverman) [inline kmeans-sklearn/README.md]: Should be right after 'visit' sentence

File: kmeans-sklearn/kmeans_image_quant.py [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+    num_colors = int(query['num_colors']) if ('num_colors' in query) else default_number_colors
+
+    extension = os.path.splitext(image_url)[1]
+    image_data = compress_image(image_url, num_colors)
+    image_path = default_file_path % extension
+    save_image(image_data, image_path)
+
+    # our image needs to be in the correct format to return it over HTTP
+    with open(image_path, 'rb') as imageFile:
+        compressed_image = imageFile.read()
```
Comment (arielshaqed) [inline kmeans-sklearn/kmeans_image_quant.py]: Writing a file then reading it is a bit naff.  @bergundy 
 suggests... https://pillow.readthedocs.io/en/stable/reference/Image.html#PIL.Image.Image.save says you can save to a file object, which could be a [BytesIO](https://docs.python.org/3/library/io.html#io.BytesIO).  Now `getBuffer` the image contents.

File: kmeans-sklearn/kmeans_image_quant.py [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+def handler(body, ctx):
+    query = ctx.request.query
+    image_url = query['image_url']
+    # try and use the num_colors query param if it exists
+    num_colors = int(query['num_colors']) if ('num_colors' in query) else default_number_colors
+
+    extension = os.path.splitext(image_url)[1]
+    image_data = compress_image(image_url, num_colors)
+    image_path = default_file_path % extension
+    save_image(image_data, image_path)
```
Comment (arielshaqed) [inline kmeans-sklearn/kmeans_image_quant.py]: Using this as the output format doesn't work if I pass a JPEG (or any lossy compression format).  It will be even more visible once you start using e.g. Floyd-Steinberg (JPEG _hates_ dithering).
Instead output should be e.g. image/png.

File: kmeans-sklearn/README.md (hunk 2 above)
Comment (arielshaqed) [reply kmeans-sklearn/README.md]: In any case, "it is assumed" is academic and a bit stilted.  I'd prefer e.g. "You should already have a Binaris account".

Comment (rylandg) [reply kmeans-sklearn/README.md]: Do you like the new version?

File: kmeans-sklearn/kmeans_image_quant.py (hunk 7 above)
Comment (rylandg) [reply kmeans-sklearn/kmeans_image_quant.py]: So you don't think image type should be preserved?

**Comment by rylandg** [comment]
@arielshaqed done. 

I tested both locally and in the function and it works great using the BufferIO change. Thank Roey for me :)

File: kmeans-sklearn/README.md [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+
+As a final note, keep in mind that you only need to run Bob again if you add or remove pip packages. Changes to the function code itself only require another `bn deploy`.
+
+## Try it out
+
+Visit the URL printed by deploy (for `public_kmeans_image_quant_py2`) in your browser and add the image URL you want to run kmeans on.
+
+_Example URL_
+
+https://run.binaris.com/v2/run/0123456789/public_kmeans_image_quant_py2?image_url=https://i.imgur.com/HLmmVIg.jpg
```
Comment (michaeladda) [inline kmeans-sklearn/README.md]: 1. This link (and the one below) are clickable yet generate an error. Very misleading. Please either add an account for the functions examples and make it live, or render the link unclickable somehow. It otherwise fosters frustration.

1. 1. Might also want to link directly to the original picture and a pre-generated result for comparison
   1. Might choose a different picture (this one was picked I suspect due to color scheme). Maybe Binaris related suitable picture?

File: kmeans-sklearn/kmeans_image_quant.py [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+import urllib2, cStringIO
+import io
+
+from PIL import Image
+from sklearn.cluster import KMeans
+from sklearn.metrics import pairwise_distances_argmin
+from sklearn.datasets import load_sample_image
+from sklearn.utils import shuffle
+
+# This function utilizes a modified version of the scikit-learn example available here
```
Comment (michaeladda) [inline kmeans-sklearn/kmeans_image_quant.py]: not really utilizes, right? from reading the link it is either based on or inspired by?

File: kmeans-sklearn/kmeans_image_quant.py [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+    labels = kmeans.predict(image_array)
+
+    # use our output clustering to reconstruct the image with a reduce pallete
+    return recreate_image(kmeans.cluster_centers_, labels, w, h)
+
+def handler(body, ctx):
+    query = ctx.request.query
+    image_url = query['image_url']
+    # try and use the num_colors query param if it exists
+    num_colors = int(query['num_colors']) if ('num_colors' in query) else default_number_colors
```
Comment (michaeladda) [inline kmeans-sklearn/kmeans_image_quant.py]: maybe default_num_colors to keep them similar?
also, consider validating the input (url & num)

File: kmeans-sklearn/README.md [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+
+If this is your first time using Binaris, visit our [Getting Started](https://dev.binaris.com/tutorials/python/getting-started/) page to set up your account for free.
+
+There are two functions in this example
+
+#### public_kmeans_image_quant_py2
+
+Uses kmeans to quantize an input image potentially reducing the number of unique colors greatly. This can allow for greater compression of the input image or pleasing filter-esque effects.
+
+#### public_kmeans_simple_py2
```
Comment (michaeladda) [inline kmeans-sklearn/README.md]: what's the purpose of this example?
iiuc it does nothing. generates the model  during initialization and prints the same labels over and over when invoked. why not use it to identify an iris?

Comment (rylandg) [reply kmeans-sklearn/README.md]: This is meant to be an extremely simple example which just proves/shows a running kmeans prediction on the system. I chose to do it at initialization time so latency isn't effected outside of the first run.

Identifying an Iris means you know how to properly structure a query with Iris data. Would you rather I

1. Add a simple script that invokes the function with a valid 'iris' in the body
2. Remove the example altogether.

File: kmeans-sklearn/README.md (hunk 8 above)
Comment (rylandg) [reply kmeans-sklearn/README.md]: 1. Sure, i'll make it unclickable
2. i. sure
     ii. let me see what I can do

File: kmeans-sklearn/kmeans_image_quant.py (hunk 9 above)
Comment (rylandg) [reply kmeans-sklearn/kmeans_image_quant.py]: inspired is a better choice, will fix.

File: kmeans-sklearn/kmeans_image_quant.py (hunk 10 above)
Comment (rylandg) [reply kmeans-sklearn/kmeans_image_quant.py]: 1. Will make them uniform
2. Not sure, it's adds more code and I'm not sure people assume it's battle hardened.

File: kmeans-sklearn/README.md [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+ht&#8203;tps://run.binaris.com/v2/run/0123456789/public_kmeans_image_quant_py2?image_url=https://i.imgur.com/DUSZHiX.jpg&num_colors=16
+
+
+### Example output
+
+Below are example before and after images after running the function with `num_colors=2`
+
+_Before_
+
+![before_kmeans](./res/before_kmeans.jpg)
```
Comment (michaeladda) [inline kmeans-sklearn/README.md]: the images are missing

File: kmeans-sklearn/README.md [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+
+As a final note, keep in mind that you only need to run Bob again if you add or remove pip packages. Changes to the function code itself only require another `bn deploy`.
+
+## Try it out
+
+Visit the URL printed by deploy (for `public_kmeans_image_quant_py2`) in your browser and add the image URL you want to run kmeans on.
+
+_Example URL_
+
+ht&#8203;tps://run.binaris.com/v2/run/0123456789/public_kmeans_image_quant_py2?image_url=https://i.imgur.com/DUSZHiX.jpg
```
Comment (michaeladda) [inline kmeans-sklearn/README.md]: maybe hint to replace 0123456789 with the account id?

File: kmeans-sklearn/README.md [hunk 14]
```diff
@@ -0,0 +1,63 @@
+# KMeans via sklearn (Python2)
+
+Two clustering examples using [sklearn](https://scikit-learn.org/stable/).
```
Comment (michaeladda) [inline kmeans-sklearn/README.md]: One now?

**Comment by rylandg** [comment]
@michaeladda added images, changed `two` to `one` and added note about account id
